lib/within.js

```diff
@@ -21,19 +21,68 @@ function within(context, fn) {
       if (helpers[helper]._withinBegin) recorder.add('start within block', () => helpers[helper]._withinBegin(context));
     });
 
-    fn();
 
-    Object.keys(helpers).forEach((helper) => {
-      if (helpers[helper]._withinEnd) recorder.add('finish within block', () => helpers[helper]._withinEnd());
-    });
-
-    recorder.add('restore session', () => {
-      recorder.session.restore();
+    try {
```

Comment:
this is not fully copy-paste. There are several changes. There are also changes in recorder.catch() for within (we have to pass error as it appears in within to stop promises chain in main generator)